### PR TITLE
HPCC-21723 Stop persistent handler thread at shutdown

### PR DESCRIPTION
**Pull request by mayx**
- Add MODULE_INIT adn MODULE_EXIT, in MODULE_EXIT clear up the static
  http context variable
- Stop persistent handler explicitly in http context's destructor

Signed-off-by: mayx <yanrui.ma@lexisnexisrisk.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by mayx** [comment]
@afishbeck please review...

**Comment by richardkchapman** [comment]
@ghalliday Can you review since Tony is away

File: esp/bindings/http/client/httpclient.cpp [hunk 1]
```diff
@@ -1001,3 +1008,7 @@ IHttpClientContext* createHttpClientContext(IPropertyTree* config)
     return new CHttpClientContext(config);
 }
 
+MODULE_EXIT()
```
Comment (afishbeck) [inline esp/bindings/http/client/httpclient.cpp]: Trivial, but the convention seems to be to put the MODULE_EXIT() at the top right after the MODULE_INIT().  It's easy to keep them in sync that way.

**Review by afishbeck — CHANGES_REQUESTED** [review]
@mayx one trivial change, otherwise looks goo.

File: esp/bindings/http/client/httpclient.cpp (hunk 1 above)
Comment (mayx) [reply esp/bindings/http/client/httpclient.cpp]: Change made as suggested.

**Comment by mayx** [comment]
@afishbeck please review again...

File: esp/bindings/http/client/httpclient.cpp [hunk 2]
```diff
@@ -29,6 +29,19 @@
...
+static CriticalSection httpCrit;
+
+MODULE_INIT(INIT_PRIORITY_STANDARD)
+{
+    return true;
+}
+
+MODULE_EXIT()
+{
+	theHttpClientContext.clear();
```
Comment (afishbeck) [inline esp/bindings/http/client/httpclient.cpp]: Extra indent?

File: esp/bindings/http/client/httpclient.cpp [hunk 3]
```diff
@@ -68,6 +81,8 @@ void CHttpClientContext::initPersistentHandler()
 
 CHttpClientContext::~CHttpClientContext()
 {
+	if (m_persistentHandler.get())
+		m_persistentHandler->stop(true);
```
Comment (afishbeck) [inline esp/bindings/http/client/httpclient.cpp]: Extra indents?

**Review by afishbeck — CHANGES_REQUESTED** [review]
Sorry, just noticed the indents.

**Comment by mayx** [comment]
@afishbeck indent fixed please review again. Just realized I forgot to add the checks in gitconfig after reinstalling my box, so tabs crept in...

**Review by afishbeck — APPROVED** [review]
@mayx looks good.

**Pull request merged**